File: Shundori/nodejs/reports.js
/**
 * Report generation and CSV export functionality (Node.js)
 */
const { pool } = require('./database');
const { getPhnomPenhDate, config } = require('./config');
const { getFineAmount } = require('./attendance');
const createCsvWriter = require('csv-writer').createObjectCsvWriter;
const fs = require('fs');
const path = require('path');

function formatUserName(user) {
  if (user.full_name) {
    return user.full_name;
  } else if (user.username) {
    return `@${user.username}`;
  } else {
    return `User ${user.telegram_id}`;
  }
}

async function generateDailyReport(reportDate = null) {
  if (!reportDate) {
    reportDate = getPhnomPenhDate();
  }

  const client = await pool.connect();
  try {
    // Get all active users
    const usersResult = await client.query('SELECT * FROM users WHERE is_active = TRUE');
    const users = usersResult.rows;

    const presentUsers = [];
    const absentUsers = [];
    const fineAmount = await getFineAmount();

    for (const user of users) {
      // Check attendance record for the date
      const recordResult = await client.query(
        'SELECT * FROM attendance_records WHERE user_id = $1 AND date = $2',
        [user.id, reportDate]
      );

      const record = recordResult.rows[0];

      if (record && record.status === 'present') {
        presentUsers.push({
          user: user,
          timestamp: record.timestamp
        });
      } else {
        // Check if fine was already recorded
        const fineResult = await client.query(
          'SELECT * FROM fines WHERE user_id = $1 AND date = $2',
          [user.id, reportDate]
        );

        const fine = fineResult.rows[0];
        absentUsers.push({
          user: user,
          fine: fine ? fine.amount : fineAmount
        });
      }
    }

    // Calculate running fines
    const runningFines = {};
    for (const user of users) {
      const finesResult = await client.query(
        'SELECT * FROM fines WHERE user_id = $1',
        [user.id]
      );
      runningFines[user.id] = finesResult.rows.reduce((sum, f) => sum + f.amount, 0);
    }

    return {
      date: reportDate,
      total_members: users.length,
      present: presentUsers,
      absent: absentUsers,
      running_fines: runningFines,
      all_users: users
    };
  } finally {
    client.release();
  }
}

function formatDailyReportMessage(report, includeRunningFines = false) {
  const dateStr = report.date;
  let message = `📊 Daily Attendance Report - ${dateStr}\n\n`;
  message += `👥 Total Members: ${report.total_members}\n\n`;

  // Present members
  message += '✅ Present Members:\n';
  if (report.present.length > 0) {
    for (const item of report.present) {
      const userName = formatUserName(item.user);
      const timestamp = item.timestamp
        ? new Date(item.timestamp).toLocaleTimeString('en-US', { hour12: false, timeZone: 'Asia/Phnom_Penh' })
        : 'N/A';
      message += `  • ${userName} (${timestamp})\n`;
    }
  } else {
    message += '  None\n';
  }

  message += '\n';

  // Absent/Late members
  message += '❌ Absent/Late Members:\n';
  if (report.absent.length > 0) {
    for (const item of report.absent) {
      const userName = formatUserName(item.user);
      const fine = item.fine;
      message += `  • ${userName} - Fine: $${fine.toFixed(2)}\n`;
    }
  } else {
    message += '  None\n';
  }

  if (includeRunningFines) {
    message += '\n💰 Running Fines:\n';
    for (const [userId, total] of Object.entries(report.running_fines)) {
      if (total > 0) {
        const user = report.all_users.find(u => u.id === parseInt(userId));
        if (user) {
          const userName = formatUserName(user);
          message += `  • ${userName}: $${total.toFixed(2)}\n`;
        }
      }
    }
  }

  return message;
}

async function exportDailyCSV(reportDate = null, outputDir = 'exports') {
  if (!reportDate) {
    reportDate = getPhnomPenhDate();
  }

  // Create exports directory
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  const client = await pool.connect();
  try {
    const usersResult = await client.query('SELECT * FROM users WHERE is_active = TRUE');
    const users = usersResult.rows;
    const fineAmount = await getFineAmount();

    const records = [];
    for (const user of users) {
      const recordResult = await client.query(
        'SELECT * FROM attendance_records WHERE user_id = $1 AND date = $2',
        [user.id, reportDate]
      );

      const fineResult = await client.query(
        'SELECT * FROM fines WHERE user_id = $1 AND date = $2',
        [user.id, reportDate]
      );

      const record = recordResult.rows[0];
      const fine = fineResult.rows[0];

      records.push({
        Date: reportDate,
        'Telegram ID': user.telegram_id,
        Username: user.username || '',
        'Full Name': user.full_name || '',
        Status: record ? record.status : 'absent',
        Timestamp: record && record.timestamp
          ? new Date(record.timestamp).toISOString()
          : '',
        'Fine Amount': fine ? fine.amount : (record && record.status === 'present' ? 0 : fineAmount)
      });
    }

    const filename = `attendance_${reportDate.replace(/-/g, '')}.csv`;
    const filepath = path.join(outputDir, filename);

    const csvWriter = createCsvWriter({
      path: filepath,
      header: [
        { id: 'Date', title: 'Date' },
        { id: 'Telegram ID', title: 'Telegram ID' },
        { id: 'Username', title: 'Username' },
        { id: 'Full Name', title: 'Full Name' },
        { id: 'Status', title: 'Status' },
        { id: 'Timestamp', title: 'Timestamp' },
        { id: 'Fine Amount', title: 'Fine Amount' }
      ]
    });

    await csvWriter.writeRecords(records);
    return filepath;
  } finally {
    client.release();
  }
}

async function exportMonthlyCSV(year, month, outputDir = 'exports') {
  const { pool } = require('./database');
  const client = await pool.connect();

  try {
    // Create exports directory
    if (!fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir, { recursive: true });
    }

    const startDate = `${year}-${month.toString().padStart(2, '0')}-01`;
    const lastDay = new Date(year, month, 0).getDate();
    const endDate = `${year}-${month.toString().padStart(2, '0')}-${lastDay}`;

    const usersResult = await client.query('SELECT * FROM users WHERE is_active = TRUE');
    const users = usersResult.rows;
    const fineAmount = await getFineAmount();

    const records = [];
    for (const user of users) {
      // Get all attendance records for the month
      const attendanceResult = await client.query(
        'SELECT * FROM attendance_records WHERE user_id = $1 AND date >= $2 AND date <= $3',
        [user.id, startDate, endDate]
      );

      // Get all fines for the month
      const finesResult = await client.query(
        'SELECT * FROM fines WHERE user_id = $1 AND date >= $2 AND date <= $3',
        [user.id, startDate, endDate]
      );

      // Create a set of dates with attendance
      const presentDates = new Set();
      attendanceResult.rows.forEach(r => {
        if (r.status === 'present') {
          presentDates.add(r.date.toISOString().split('T')[0]);
        }
      });

      // Calculate totals
      const totalPresent = presentDates.size;
      const totalDays = lastDay;
      const totalAbsent = totalDays - totalPresent;
      const totalFines = finesResult.rows.reduce((sum, f) => sum + f.amount, 0);

      records.push({
        'Telegram ID': user.telegram_id,
        Username: user.username || '',
        'Full Name': user.full_name || '',
        'Total Present': totalPresent,
        'Total Absent': totalAbsent,
        'Total Fines': totalFines
      });
    }

    const filename = `attendance_${year}_${month.toString().padStart(2, '0')}.csv`;
    const filepath = path.join(outputDir, filename);

    const csvWriter = createCsvWriter({
      path: filepath,
      header: [
        { id: 'Telegram ID', title: 'Telegram ID' },
        { id: 'Username', title: 'Username' },
        { id: 'Full Name', title: 'Full Name' },
        { id: 'Total Present', title: 'Total Present' },
        { id: 'Total Absent', title: 'Total Absent' },
        { id: 'Total Fines', title: 'Total Fines' }
      ]
    });

    await csvWriter.writeRecords(records);
    return filepath;
  } finally {
    client.release();
  }
}

module.exports = {
  generateDailyReport,
  formatDailyReportMessage,
  exportDailyCSV,
  exportMonthlyCSV,
  formatUserName
};

File: Shundori/reports.py
"""
Report generation and CSV export functionality.
"""
from datetime import date
from typing import Dict
import pandas as pd
import os
from database import get_db, User, AttendanceRecord, Fine, Settings
from utils import format_user_name, get_phnom_penh_date
from config import DEFAULT_FINE_AMOUNT


def get_fine_amount() -> float:
    """Get current fine amount from settings."""
    with get_db() as db:
        setting = db.query(Settings).filter(Settings.key == 'fine_amount').first()
        if setting:
            return float(setting.value)
    return DEFAULT_FINE_AMOUNT


def generate_daily_report(report_date: date = None) -> Dict:
    """Generate daily attendance report."""
    if report_date is None:
        report_date = get_phnom_penh_date()
    
    with get_db() as db:
        # Get all active users
        users = db.query(User).filter(User.is_active == True).all()
        
        present_users = []
        absent_users = []
        fine_amount = get_fine_amount()
        
        for user in users:
            # Check attendance record for the date
            record = db.query(AttendanceRecord).filter(
                AttendanceRecord.user_id == user.id,
                AttendanceRecord.date == report_date
            ).first()
            
            if record and record.status == 'present':
                present_users.append({
                    'user': user,
                    'timestamp': record.timestamp
                })
            else:
                # Check if fine was already recorded
                fine = db.query(Fine).filter(
                    Fine.user_id == user.id,
                    Fine.date == report_date
                ).first()
                
                fine_amount_for_user = fine_amount
                if fine:
                    fine_amount_for_user = fine.amount
                elif record and record.status == 'late':
                    # Late users also get fined
                    fine_amount_for_user = fine_amount
                
                absent_users.append({
                    'user': user,
                    'fine': fine_amount_for_user
                })
        
        # Calculate running fines
        running_fines = {}
        for user in users:
            total_fines = db.query(Fine).filter(Fine.user_id == user.id).all()
            running_fines[user.id] = sum(f.amount for f in total_fines)
        
        return {
            'date': report_date,
            'total_members': len(users),
            'present': present_users,
            'absent': absent_users,
            'running_fines': running_fines,
            'all_users': users
        }


def format_daily_report_message(report: Dict, include_running_fines: bool = False) -> str:
    """Format daily report as a message."""
    date_str = report['date'].strftime('%Y-%m-%d')
    message = f"📊 Daily Attendance Report - {date_str}\n\n"
    message += f"👥 Total Members: {report['total_members']}\n\n"
    
    # Present members
    message += "✅ Present Members:\n"
    if report['present']:
        for item in report['present']:
            user_name = format_user_name(item['user'])
            timestamp_str = 'N/A'
            if item.get('timestamp'):
                try:
                    timestamp_str = item['timestamp'].strftime('%H:%M:%S')
                except (AttributeError, TypeError):
                    timestamp_str = 'N/A'
            message += f"  • {user_name} ({timestamp_str})\n"
    else:
        message += "  None\n"
    
    message += "\n"
    
    # Absent/Late members
    message += "❌ Absent/Late Members:\n"
    if report['absent']:
        for item in report['absent']:
            user_name = format_user_name(item['user'])
            fine = item['fine']
            message += f"  • {user_name} - Fine: ${fine:.2f}\n"
    else:
        message += "  None\n"
    
    if include_running_fines:
        message += "\n💰 Running Fines:\n"
        for user_id, total in report['running_fines'].items():
            if total > 0:
                # Find user
                user = next((u for u in report.get('all_users', []) if u.id == user_id), None)
                if user:
                    user_name = format_user_name(user)
                    message += f"  • {user_name}: ${total:.2f}\n"
    
    return message


def export_daily_csv(report_date: date = None, output_dir: str = 'exports') -> str:
    """Export daily report to CSV file."""
    if report_date is None:
        report_date = get_phnom_penh_date()
    
    os.makedirs(output_dir, exist_ok=True)
    
    with get_db() as db:
        users = db.query(User).filter(User.is_active == True).all()
        fine_amount = get_fine_amount()
        
        records = []
        for user in users:
            record = db.query(AttendanceRecord).filter(
                AttendanceRecord.user_id == user.id,
                AttendanceRecord.date == report_date
            ).first()
            
            fine = db.query(Fine).filter(
                Fine.user_id == user.id,
                Fine.date == report_date
            ).first()
            
            records.append({
                'Date': report_date.strftime('%Y-%m-%d'),
                'Telegram ID': user.telegram_id,
                'Username': user.username or '',
                'Full Name': user.full_name or '',
                'Status': record.status if record else 'absent',
                'Timestamp': record.timestamp.strftime('%Y-%m-%d %H:%M:%S') if record and record.timestamp else '',
                'Fine Amount': fine.amount if fine else (fine_amount if not record or record.status != 'present' else 0)
            })
    
    df = pd.DataFrame(records)
    filename = f"attendance_{report_date.strftime('%Y%m%d')}.csv"
    filepath = os.path.join(output_dir, filename)
    df.to_csv(filepath, index=False)
    
    return filepath


def export_monthly_csv(year: int, month: int, output_dir: str = 'exports') -> str:
    """Export monthly report to CSV file."""
    from calendar import monthrange
    
    os.makedirs(output_dir, exist_ok=True)
    
    start_date = date(year, month, 1)
    end_date = date(year, month, monthrange(year, month)[1])
    
    with get_db() as db:
        users = db.query(User).filter(User.is_active == True).all()
        fine_amount = get_fine_amount()
        
        records = []
        for user in users:
            # Get all attendance records for the month
            attendance_records = db.query(AttendanceRecord).filter(
                AttendanceRecord.user_id == user.id,
                AttendanceRecord.date >= start_date,
                AttendanceRecord.date <= end_date
            ).all()
            
            # Get all fines for the month
            fines = db.query(Fine).filter(
                Fine.user_id == user.id,
                Fine.date >= start_date,
                Fine.date <= end_date
            ).all()
            
            # Create a set of dates with attendance
            present_dates = {r.date for r in attendance_records if r.status == 'present'}
            
            # Calculate totals
            total_present = len(present_dates)
            total_absent = (end_date - start_date).days + 1 - total_present
            total_fines = sum(f.amount for f in fines)
            
            records.append({
                'Telegram ID': user.telegram_id,
                'Username': user.username or '',
                'Full Name': user.full_name or '',
                'Total Present': total_present,
                'Total Absent': total_absent,
                'Total Fines': total_fines
            })
    
    df = pd.DataFrame(records)
    filename = f"attendance_{year:04d}_{month:02d}.csv"
    filepath = os.path.join(output_dir, filename)
    df.to_csv(filepath, index=False)
    
    return filepath

